### PR TITLE
Shaunna Wiens --carets

### PR DESCRIPTION
**Pull request by skwiens**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | One design decision I had to make when working on this project is which classes to make. One option I was considering was making two classes: reservations and rooms. In the end I decided to make a hotel class as well as Pricing class. If I were to do it again, I wouldn't include the pricing class, but it seemed as though it would be the same amount of work to move it now or later, so according to the idea in the book I left it where it was.  Ultimately, I think the hotel has too many responsibilities according to good OOD, but it helped keep the other classes simpler.       |
| Describe a concept that you gained more clarity on as you worked on this assignment. | As I worked on this assignment I gained more and less clarity on testing.  I gained more clarity on red, green, refactor as a process and how to write in that order.  I gained less clarity (or rather ran into more difficulties) as to how to test certain aspects of my program. I gave up on testing the CLI through tests and took it's spec out of the loop by changing the file name.    |
| Describe a _nominal test_ that you wrote for this assignment. |  A nominal test that I wrote for this assignment was the ability to access the attr_readers correctly.   |
| Describe an _edge case test_ that you wrote for this assignment. |  Some edge case tests that I wrote were to not allow a user to use invalid dates such as checking in before today or checking out the day of the check in or before.    |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | Most of the times I wrote out tests and then code. I don't think I did great at writing out pseudocode, but many times I would write the description portion of my tests followed by a test and then code to match. The description of my test is somewhat like pseudocode.   |

### CODE REVIEW

File: lib/hotel.rb [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        res_by_date << res
+      end
+    end
+    return res_by_date
+  end
+
+  def get_available_rooms(date_begin, date_end)
+    available_rooms = @rooms.clone
+
+    @reservations.each do |res|
```
Comment (CheezItMan) [inline lib/hotel.rb]: This type of loop looks repeated, maybe you could turn that into a helper method.

File: lib/hotel.rb [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  # def find_room_by_number(room_number)
+  #   rooms.each do |room|
+  #     if room.number == room_number
+  #       return room
+  #     end
+  #   end
+  #   raise StandardError.new "this room doesn't exist"
+  # end
+
+  def assign_id
```
Comment (CheezItMan) [inline lib/hotel.rb:154]: I'm not sure what purpose this method serves, why make the block Ids random?

File: specs/hotel_spec.rb [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    end
+
+    it "creates 20 rooms that are room objects" do
+      marriott = Hotel::Hotel.new
+      marriott.rooms.each do |room|
+        room.must_be_instance_of Hotel::Room
+      end
+    end
+
+    it "creates 20 hotels that can identify their room numbers" do
```
Comment (CheezItMan) [inline specs/hotel_spec.rb:23]: This is probably too specific as to the implementation of the rooms, like that they are in order.  Maybe just test that `rooms` must include a room with each id.

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| Good number of commit and good messages. |
Answer comprehension questions	|  Check, I agree with you about `Pricing` and `Hotel`.   Testing the CLI isn't something you can do without using a Mocking tool, which we won't use for now.     |
**Design** |
Each class is responsible for a single piece of the program | Pretty good separation, I like the reservation class with an `include_date?` method.  The `Pricing` class looks more like a method for a room instead of a class by itself. |
Classes are loosely coupled | Fairly loosely coupled, but `Hotel` is doing a huge amount of the work. |
**Wave 1** |
List rooms |  Check, well done |
Reserve a room for a given date range | Check  |
List reservations for a given date | Check  |
Calculate reservation price | Check, pricing class |
Invalid date range produces an error | Well done, methods to produce the checks make things a bit more readable. |
**Wave 2** |
View available rooms for a given date range | Check | 
Reserving a room that is not available produces an error | Check  |
**Wave 3** |
Create a block of rooms | Interesting choice that the block includes a list of rooms.   |
Check if a block has rooms | Check via `rooms_in_block`  |
Reserve a room from a block | Check | 
Test coverage |  99.75%!! |
**Additional Feedback** | Nicely done!  In your testing for `Block` you should include testing for things such as invalid dates at the `Block` class level because the class might be reused elsewhere.  Also in your testing, what about when you try to reserve a room with an invalid room #?  Good edge case testing to see if you can reserve a room on the day someone is checking out.  For testing the uniqueness of the block booking you can make a hash of all the booking IDs and loop through counting how often they occur as the value.  Then if all the values are 1, they're unique.  Because they're random there's no way to verify it 100%, but you could run the test several times.  For `returns 17 rooms when there are three room conflicts` you can also just verify that the available rooms don't include the rooms in the specific reservations.  Overall you did really really well.  The design was good, although as you noted, not perfect.  You also did a very through job testing.    |

**Review by CheezItMan — COMMENTED** [review]
Nicely done, just a few in-code comments.
